File: 015_Binary_Search_Advanced_Problems/Book_Allocation_Problem.swift
class Solution {
    func splitArray(_ array: [Int], _ key: Int) -> Int {
        func requiredSplit(_ largest: Int) -> Int {
            var currSum = 0
            var splitCount = 0

            for element in array {
                if currSum + element <= largest {
                    currSum += element
                } else {
                    currSum = element
                    splitCount += 1
                }
            }
            return splitCount + 1
        }

        var left = array.max() ?? 0
        var right = array.reduce(0, +)
        var answer = right

        while left <= right {
            let mid = left + (right - left) / 2
            if requiredSplit(mid) <= key {
                answer = mid
                right = mid - 1
            } else {
                left = mid + 1
            }
        }
        return answer
    }
}

// Example usage
let arr = readLine()!.split(separator: " ").compactMap { Int($0) }
let m = Int(readLine()!)!
let ob = Solution()
print(ob.splitArray(arr, m))

File: 015_Binary_Search_Advanced_Problems/Book_Allocation_Problem.py
class Solution:
    def splitArray(self, Array: [int], Key: int) -> int:
        def required_split(largest):

            curr_sum  = 0
            split_count = 0

            for Element in Array:
                if curr_sum + Element <= largest:
                    curr_sum += Element
                else:
                    curr_sum = Element
                    split_count += 1
            return split_count + 1

        Left = max(Array)
        Right = sum(Array)
        Answer = Right

        while Left <= Right:
            mid = Left + ((Right-Left)// 2)
            if required_split(mid) <= Key:
                Answer = mid
                Right = mid - 1
            else:
                Left = mid + 1
        return Answer

if __name__=='__main__':     
    arr=[int(x) for x in input().strip().split()]
    m=int(input())        
    ob=Solution()        
    print(ob.splitArray(arr,m))